File: source/GyeongHo/boj/1463.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int cache[1000001];
int N;

int main()
{
	cin >> N;
	cache[1] = 0;
	cache[2] = 1;

	for(int i = 3; i < N + 1; ++i) {
		cache[i] = 1 + cache[i - 1];
		if(i % 3 == 0) cache[i] = min(cache[i/3] + 1, cache[i]);
		if(i % 2 == 0) cache[i] = min(cache[i/2] + 1, cache[i]);
	}

	cout << cache[N];
	return 0;
}

File: source/GyeongHo/boj/1463.cs
using System;

namespace evaluateMIN
{
	class makeInOne
	{
		static int[] cache = new int[1000001];

		static int dfs(int x)
		{
			if(cache[x] != -1) return cache[x];
			else if(x == 1) return cache[x] = 0;
			else if(x % 3 == 0) return cache[x] = 1 + dfs(x / 3);
			else if(x % 2 == 0) return cache[x] = 1 + dfs(x / 2);
			else return cache[x] = 1 + dfs(x - 1);
		}

		static void Main()
		{
			for(int i = 0; i < 1000001; ++i)
				cache[i] = -1;

			int x;
			x = int.Parse(Console.ReadLine());

			int MIN = dfs(x);
			for(int i = 0; i < x; ++i) {
				int tmp = i + dfs(x - i);
				if(MIN > tmp)
					MIN = tmp;
			}

			Console.Write(MIN);
		}
	}
}
